# Dummy Variables Exercise

In this exercise, you'll create dummy variables from the projects data set. The idea is to transform categorical data like this:

| Project ID | Project Category |
|------------|------------------|
| 0          | Energy           |
| 1          | Transportation   |
| 2          | Health           |
| 3          | Employment       |

into new features that look like this:

| Project ID | Energy | Transportation | Health | Employment |
|------------|--------|----------------|--------|------------|
| 0          | 1      | 0              | 0      | 0          |
| 1          | 0      | 1              | 0      | 0          |
| 2          | 0      | 0              | 1      | 0          |
| 3          | 0      | 0              | 0      | 1          |


(Note if you were going to use this data with a model influenced by multicollinearity, you would want to eliminate one of the columns to avoid redundant information.) 

The reasoning behind these transformations is that machine learning algorithms read in numbers not text. Text needs to be converted into numbers. You could assign a number to each category like 1, 2, 3, and 4. But a categorical variable has no inherent order, so you want to reflect this in your features.

Pandas makes it very easy to create dummy variables with the [get_dummies](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.get_dummies.html) method. In this exercise, you'll create dummy variables from the World Bank projects data; however, there's a caveat. The World Bank data is not particularly clean, so you'll need to explore and wrangle the data first.

You'll focus on the text values in the sector variables.

Run the code cells below to read in the World Bank projects data set and then to filter out the data for text variables. 

In [11]:
import pandas as pd
import numpy as np

# read in the projects data set and do basic wrangling 
projects = pd.read_csv('../data/projects_data.csv', dtype=str)
projects.drop('Unnamed: 56', axis=1, inplace=True)
projects['totalamt'] = pd.to_numeric(projects['totalamt'].str.replace(',', ''))
projects['countryname'] = projects['countryname'].str.split(';', expand=True)[0]
projects['boardapprovaldate'] = pd.to_datetime(projects['boardapprovaldate'])

# keep the project name, lending, sector and theme data
sector = projects.copy()
sector = sector[['project_name', 'lendinginstr', 'sector1', 'sector2', 'sector3', 'sector4', 'sector5', 'sector',
          'mjsector1', 'mjsector2', 'mjsector3', 'mjsector4', 'mjsector5',
          'mjsector', 'theme1', 'theme2', 'theme3', 'theme4', 'theme5', 'theme ',
          'goal', 'financier', 'mjtheme1name', 'mjtheme2name', 'mjtheme3name',
          'mjtheme4name', 'mjtheme5name']]

Run the code cell below. This cell shows the percentage of each variable that is null. Notice the mjsector1 through mjsector5 variables are all null. The mjtheme1name through mjtheme5name are also all null as well as the theme variable. 

Because these variables contain so many null values, they're probably not very useful.

In [12]:
# output percentage of values that are missing
100 * sector.isnull().sum() / sector.shape[0]

project_name      0.000000
lendinginstr      1.348093
sector1           0.000000
sector2          47.791539
sector3          64.450899
sector4          76.019290
sector5          85.132617
sector            3.496274
mjsector1       100.000000
mjsector2       100.000000
mjsector3       100.000000
mjsector4       100.000000
mjsector5       100.000000
mjsector          3.496274
theme1            0.000000
theme2           46.005042
theme3           58.987286
theme4           71.317405
theme5           83.954406
theme           100.000000
goal             33.510522
financier        61.310829
mjtheme1name    100.000000
mjtheme2name    100.000000
mjtheme3name    100.000000
mjtheme4name    100.000000
mjtheme5name    100.000000
dtype: float64

# Exercise 1

The sector1 variable looks promising; it doesn't contain any null values at all. In the next cell, store the unique sector1 values in a list and output the results. Use the sort_values() and unique() methods.

In [13]:
# TODO: Create a list of the unique values in sector1. Use the sort_values() and unique() pandas methods. 
# And then convert those results into a Python list
uniquesectors1 = sector['sector1'].unique()
uniquesectors1

array(['!$!0', 'Social Protection!$!63!$!SA', 'Urban Transport!$!95!$!TC',
       ..., 'Sewerage!$!79!$!WS',
       'Public Administration - Information and Communications Technologies!$!3!$!CF',
       '(Historic)HIV/AIDS!$!100!$!HA'], dtype=object)

In [14]:
# run this code cell to see the number of unique values
print('Number of unique values in sector1:', len(uniquesectors1))

Number of unique values in sector1: 3060


3060 different categories is quite a lot! Remember that with dummy variables, if you have n categorical values, you need n - 1 new variables! That means 3059 extra columns! 

# Exercise 2

There are a few issues with this 'sector1' variable. First, there are values labeled '!$!0'. These should be substituted with NaN.

Furthermore, each sector1 value ends with a ten or eleven character string like '!$!49!$!EP'. Some sectors show up twice in the list like:
 'Other Industry; Trade and Services!$!70!$!YZ',
 'Other Industry; Trade and Services!$!63!$!YZ',

But it seems like those are actually the same sector. You'll need to remove everything past the exclamation point. 

Many values in the sector1 variable start with the term '(Historic)'. Try removing that phrase as well.

### replace() method

With pandas, you can use the replace() method to search for text and replace parts of a string with another string. If you know the exact string you're looking for, the replace() method is straight forward. For example, say you wanted to remove the string '(Trial)' from this data:

| data                     |
|--------------------------|
| '(Trial) Banking'        ||
| 'Banking'                |
| 'Farming'                |
| '(Trial) Transportation' |

You could use `df['data'].replace('(Trial'), '')` to replace (Trial) with an empty string.

### regular expressions
What about this data?

| data                                           |
|------------------------------------------------|
| 'Other Industry; Trade and Services?$ab' |
| 'Other Industry; Trade and Services?ceg' |

This type of data is trickier. In this case, there's a pattern where you want to remove a string that starts with an exclamation point and then has an unknown number of characters after it. When you need to match patterns of character, you can use [regular expressions](https://en.wikipedia.org/wiki/Regular_expression).

The replace method can take a regular expression. So
df['data'].replace('?.+', regex=True) where '?.+' means find a set of characters that starts with a question mark is then followed by one or more characters. You can see a [regular expression cheat sheet](https://medium.com/factory-mind/regex-tutorial-a-simple-cheatsheet-by-examples-649dc1c3f285) here.

Fix these issues in the code cell below.

In [15]:
# TODO: In the sector1 variable, replace the string '!$10' with nan
#       Put the results back into the sector1 variable
# HINT: you can use the pandas replace() method and numpy.nan
sector['sector1'].replace('!$!0', np.nan, inplace=True)

# TODO: In the sector1 variable, remove the last 10 or 11 characters from the sector1 variable.
# HINT: There is more than one way to do this. To do it with one line of code,
# you can use the replace method with a regex expression '!.+'
# That regex expression looks for a string with an exclamation
# point followed by one or more characters
sector['sector1'].replace(to_replace ='!.+' , value = '', regex = True, inplace=True)

# TODO: Remove the string '(Historic)' from the sector1 variable
# HINT: You can use the replace method
#sector['sector1'] = sector['sector1'].str.replace("(Historic)", '')
sector['sector1'].replace("^(\(Historic\))", '',regex = True, inplace=True)

print('Number of unique sectors after cleaning:', len(list(sector['sector1'].unique())))
print('Percentage of null values after cleaning:', 100 * sector['sector1'].isnull().sum() / sector['sector1'].shape[0])

Number of unique sectors after cleaning: 156
Percentage of null values after cleaning: 3.49627356423


Now there are 156 unique categorical values. That's better than 3060. If you were going to use this data with a supervised learning machine model, you could try converting these 156 values to dummy variables. You'd still have to train and test a model to see if those are good features.

You could try to consolidate similar categories together, which is what the challenge exercise in part 4 is about.

There are also still many entries with NaN values. How could you fill these in? 

You might try to determine an appropriate category from the 'project_name' or 'lendinginstr' variables. If you make dummy variables including NaN values, then you could consider a feature with all zeros to represent NaN. Or you could delete these records from the data set. Pandas will ignore NaN values by default. That means, for a given row, all dummy variables will have a value of 0 if the sector1 value was NaN.

Don't forget about the bigger context! This data is being prepared for a machine learning algorithm. Whatever techniques you use to engineer new features, you'll need to use those when running your model on new data. So if your new data does not contain a sector1 value, you'll have to run whatever feature engineering processes you did on your training set.

# Exercise 3

In this next exercise, use the pandas pd.get_dummies() method to create dummy variables. Then use the concat() method to concatenate the dummy variables to a dataframe that contains the project totalamt variable and the project year from the boardapprovaldate.

In [16]:
# TODO: Create dummy variables from the sector1 data. Put the results into a dataframe called dummies
# Hint: Use the get_dummies method
dummies = pd.get_dummies(sector['sector1'])

# TODO: Create a new dataframe called df by 
#       filtering the projects data for the totalamt and
#       the year from boardapprovaldate
projects['year'] = projects['boardapprovaldate'].dt.year
df =  projects[['totalamt','year']]

# TODO: Concatenate the results of dummies and projects
#       into a single data frame
df_final = pd.concat([df,dummies], axis=1)

df_final.head()

totalamt    year  Adult; Basic and Continuing Education  Agency reform  \
0          0  2018.0                                      0              0   
1  200000000  2018.0                                      0              0   
2   58330000  2018.0                                      0              0   
3   20000000  2018.0                                      0              0   
4  100000000  2018.0                                      0              0   

   Agricultural Extension; Research; and Other Support Activities  \
0                                                  0                
1                                                  0                
2                                                  0                
3                                                  0                
4                                                  0                

   Agricultural credit  Agricultural extension  \
0                    0                       0   
1                    0                       0   
2                    0                       0   
3                    0                       0   
4                    0                       0   

   Agricultural markets; commercialization and agri-business  \
0                                                  0           
1                                                  0           
2                                                  0           
3                                                  0           
4                                                  0           

   Agriculture adjustment  Agro-industry  \
0                       0              0   
1                       0              0   
2                       0              0   
3                       0              0   
4                       0              0   

                        ...                        Urban environment  \
0                       ...                                        0   
1                       ...                                        0   
2                       ...                                        0   
3                       ...                                        0   
4                       ...                                        0   

   Urban housing  Urban management  Urban transport  Urban water supply  \
0              0                 0                0                   0   
1              0                 0                0                   0   
2              0                 0                0                   0   
3              0                 0                0                   0   
4              0                 0                0                   0   

   Vocational training  Waste Management  Water Supply  \
0                    0                 0             0   
1                    0                 0             0   
2                    0                 0             0   
3                    0                 0             0   
4                    0                 0             0   

   Water supply and sanitation adjustment  \
0                                       0   
1                                       0   
2                                       0   
3                                       0   
4                                       0   

   Workforce Development and Vocational Education  
0                                               0  
1                                               0  
2                                               0  
3                                               0  
4                                               0  

[5 rows x 157 columns]

In [17]:
projects

id                    regionname  \
0      P162228                         Other   
1      P163962                        Africa   
2      P167672                    South Asia   
3      P158768                    South Asia   
4      P161364                        Africa   
5      P161483  Middle East and North Africa   
6      P161885                        Africa   
7      P162069                        Africa   
8      P163540                        Africa   
9      P163576  Middle East and North Africa   
10     P163721                    South Asia   
11     P164082                        Africa   
12     P164783                    South Asia   
13     P165247                        Africa   
14     P165286       Europe and Central Asia   
15     P166360  Middle East and North Africa   
16     P166734         East Asia and Pacific   
17     P167000  Middle East and North Africa   
18     P163732                        Africa   
19     P163829                        Africa   
20     P164345                        Africa   
21     P166019       Europe and Central Asia   
22     P166303       Europe and Central Asia   
23     P164429                        Africa   
24     P163756                        Africa   
25     P158756   Latin America and Caribbean   
26     P160162         East Asia and Pacific   
27     P163138         East Asia and Pacific   
28     P164262                        Africa   
29     P164289   Latin America and Caribbean   
...        ...                           ...   
18218  P000659                        Africa   
18219  P008109   Latin America and Caribbean   
18220  P037342         East Asia and Pacific   
18221  P008872       Europe and Central Asia   
18222  P008874       Europe and Central Asia   
18223  P005231  Middle East and North Africa   
18224  P006215   Latin America and Caribbean   
18225  P007499   Latin America and Caribbean   
18226  P009587                    South Asia   
18227  P007137   Latin America and Caribbean   
18228  P009133       Europe and Central Asia   
18229  P037365       Europe and Central Asia   
18230  P009586                    South Asia   
18231  P006682   Latin America and Caribbean   
18232  P009588                    South Asia   
18233  P037366       Europe and Central Asia   
18234  P037456       Europe and Central Asia   
18235  P037358       Europe and Central Asia   
18236  P006214   Latin America and Caribbean   
18237  P007497   Latin America and Caribbean   
18238  P007498   Latin America and Caribbean   
18239  P037453       Europe and Central Asia   
18240  P037454       Europe and Central Asia   
18241  P037455       Europe and Central Asia   
18242  P006577   Latin America and Caribbean   
18243  P006578   Latin America and Caribbean   
18244  P037451       Europe and Central Asia   
18245  P037362       Europe and Central Asia   
18246  P037452       Europe and Central Asia   
18247  P037383       Europe and Central Asia   

                                    countryname prodline  \
0                                         World       RE   
1              Democratic Republic of the Congo       PE   
2               People's Republic of Bangladesh       PE   
3              Islamic  Republic of Afghanistan       PE   
4                   Federal Republic of Nigeria       PE   
5                           Republic of Tunisia       PE   
6                   Federal Republic of Nigeria       PE   
7                   Federal Republic of Nigeria       PE   
8                   Federal Republic of Nigeria       PE   
9                             Lebanese Republic       PE   
10     Democratic Socialist Republic of Sri Lan       PE   
11                  Federal Republic of Nigeria       PE   
12                                        Nepal       RE   
13                  Federal Republic of Nigeria       PE   
14                              Kyrgyz Republic       RE   
15                  Hashemite Kingdom of Jordan       PE   
16             

You could continue to consolidate sector values using other techniques. For example, in the next exercise, you'll find categories with similar terms and then combine them together. 

Keep in mind that how much to consolidate will depend on your machine learning model performance and your hardware's ability to handle the extra features in memory. If your hardware's memory can handle 3060 new features and your machine learning algorithm performs better, then go for it!

# Exercise 4 (Challenge)

But can you do anything else with the sector1 variable?

The percentage of null values for 'sector1' is now 3.49%. That turns out to be the same number as the null values for the 'sector' column. You can see this if you scroll back up to where the code calculated the percentage of null values for each variable. 

Perhaps the 'sector1' and 'sector' variable have the same information. If you look at the 'sector' variable, however, it also needs cleaning. The values look like this:

'Urban Transport;Urban Transport;Public Administration - Transportation'

It turns out the 'sector' variable combines information from the 'sector1' through 'sector5' variables and the 'mjsector' variable. Run the code cell below to look at the sector variable.

In [19]:
sector['sector1']

0                                               NaN
1                                               NaN
2                                               NaN
3                                               NaN
4                                 Social Protection
5                                               NaN
6                                               NaN
7                                               NaN
8                                               NaN
9                                               NaN
10                                              NaN
11                                              NaN
12                                              NaN
13                                              NaN
14                                              NaN
15                                              NaN
16                                              NaN
17                                              NaN
18                                              NaN
19          

What else can you do? If you look at all of the diferent sector1 categories, it might be useful to combine a few of them together. For example, there are various categories with the term "Energy" in them. And then there are other categories that seem related to energy but don't have the word energy in them like "Thermal" and "Hydro". Some categories have the term "Renewable Energy", so perhaps you could make a separate "Renewable Energy" category.

Similarly, there are categories with the term "Transportation" in them, and then there are related categories like "Highways".

In the next cell, find all sector1 values with the term 'Energy' in them. For each of these rows, put the string 'energy' in a new column called 'sector1_aggregates'. Do the same for "Transportation". 

In [9]:
import re

# Create the sector1_aggregates variable
sector.loc[:,'sector1_aggregates'] = sector['sector1']

# TODO: The code above created a new variable called sector1_aggregates. 
#       Currently, sector1_aggregates has all of the same values as sector1
#       For this task, find all the rows in sector1_aggregates with the term 'Energy' in them, 
#       For all of these rows, replace whatever is the value is with the term 'Energy'.
#       The idea is to simplify the category names by combining various categories together.
#       Then, do the same for the term 'Transportation
# HINT: You can use the contains() methods. See the documentation for how to ignore case using the re library
# HINT: You might get an error saying "cannot index with vector containing NA / NaN values." 
#       Try converting NaN values to something else like False or a string

index = sector[sector.sector1_aggregates.str.contains('Energy')==True].index
sector.loc[index, 'sector1_aggregates'] = 'Energy'

index = sector[sector.sector1_aggregates.str.contains('Transportation')==True].index
sector.loc[index, 'sector1_aggregates'] = 'Transportation'

print('Number of unique sectors after cleaning:', len(list(sector['sector1_aggregates'].unique())))

Number of unique sectors after cleaning: 145


In [10]:
index = sector[sector.sector1_aggregates.str.contains('Energy')==True].index
print(index)

Int64Index([   45,    47,    57,    72,    98,   112,   145,   155,   156,
              172,
            ...
            16481, 16508, 16520, 16587, 16621, 16714, 17209, 17275, 17358,
            17386],
           dtype='int64', length=849)


# Conclusion

Pandas makes it relatively easy to create dummy variables; however, oftentimes you'll need to clean the data first.